### PR TITLE
feat(resolve): allow overriding enhanced-resolve's options

### PR DESCRIPTION
**Pull request by GertSallaerts**
Allows you to customize the config for module resolution. See: https://github.com/webpack/enhanced-resolve?tab=readme-ov-file#resolver-options

--------------

#### Old description (the PR was initially intended to only specify the `modules` option)

Allows you to do things like:

- Use bower components
- Specify your app's `src` folder as a `modules` folder and `require('models/foo');` from within your `src` folder

Basically this: https://webpack.js.org/configuration/resolve/#resolvemodules

That second bullet point used to be possible by just setting `resolvePaths` but going from [this comment](https://github.com/eslint-community/eslint-plugin-n/pull/139/files#r1389225044) on https://github.com/eslint-community/eslint-plugin-n/pull/139 that was considered an unintended side-effect or bug.

### CODE REVIEW

**Comment by scagood** [comment]
Thank you for you PR!

I will review it first thing tomorrow morning.

File: lib/util/import-target.js [hunk 1]
```diff
@@ -309,6 +310,10 @@ module.exports = class ImportTarget {
             this.resolverConfig.extensions = this.options.extensions
         }
 
+        if (this.options.modules) {
+            this.resolverConfig.modules = this.options.modules
+        }
```
Comment (scagood) [inline lib/util/import-target.js:315]: **question(non-blocking)**: I am thinking it may be best to allow all `enhanced-resolve.resolverConfig` through. 
Could the option be `this.options.resolverConfig.modules` instead of `this.options.modules`.

What do you think?

**Review by scagood — COMMENTED** [review]
Thank you for the PR!

I have one small interface question that I am curious what your thoughts are

File: lib/util/import-target.js (hunk 1 above)
Comment (GertSallaerts) [reply lib/util/import-target.js:315]: It does add some complexity on how that setting would be handled:

- Is it simply the default and are all the settings from this file (`conditionNames`, `mainFiels`, `mainFiles`, `extensions`, ...) applied over the `resolverConfig` option the consumer supplies
- Or is it an override, as in every key the consumer sets on their `resolverConfig` option will override the plugin's values. (I suspect this won't work as the plugin detects some of these based on the file that's currently being linted.)
- Or is it a mix of the two based on which key is being evaluated
- Can you pass a function that gives you the plugin's `resolverConfig` (and possible also some extra context), allowing you to adapt it to your needs (possible on a per-file basis).

If the answer is "definitively only that first option", I agree that adding `resolverConfig` as an option is way better than what I did here.

Comment (scagood) [reply lib/util/import-target.js:315]: I was thinking that anything specified by the user should overwrite all our configs. eg like this as a final step:

```js
this.resolverConfig = { ...this.resolverConfig, ...this.options.resolverConfig };
```

I think that overwriting the users config with our config would lead to confusion. That also applies to a deep merge.

The function in to mutate the config is also reasonable, but also slightly more complex that I think this warrants -- Not that I am opposed if you like the idea! :sweat_smile:

Comment (GertSallaerts) [reply lib/util/import-target.js]: Updated. I omitted the function to mutate, for now. As you mentioned: it adds complexity to the documentation, and no one may ever need it.

**Review by scagood — APPROVED** [review]
Looks good to me :+1:

File: lib/util/import-target.js (hunk 1 above)
Comment (GertSallaerts) [reply lib/util/import-target.js]: Updated the PR title and description too for future reference.

File: docs/shared-settings.md [hunk 2]
```diff
@@ -60,6 +60,26 @@ If a path is relative, it will be resolved from CWD.
 { "resolvePaths": [] }
 ```
 
+## resolverConfig
+
+Override the options generated by this plugin for resolving require and import statements. These
+options are passed down to `enhanced-resolve`'s factory method and are documented [here on their
+GitHub repository](https://github.com/webpack/enhanced-resolve?tab=readme-ov-file#resolver-options).
```
Comment (aladdin-add) [inline docs/shared-settings.md:67]: I'm a bit concerned that this will over-bind this package to `enhanced-resolve`, which may make it harder for us to make changes in the future - e.g. moving to `import.meta.resolve` (which I remember it has been discussed somewhere before).

We can state the minimum necessary in the docs. wdty?

Comment (GertSallaerts) [reply docs/shared-settings.md:67]: I guess the minimum necessary really depends on everyone's specific use cases. For us (and why I made this PR) it would be specifying  `resolverConfig.modules` to add an additional path for modules lookup. For another user it might be any of the 25 options documented in `enhanced-resolve`'s README.

Or did you mean something else by the minimum neccesary?

Comment (aladdin-add) [reply docs/shared-settings.md:67]: yes, I think `resolverConfig.modules` is enough for now. I don't feel that the implementation has to be changed as well - other options can be something like undocumented features. thoughts? /cc @scagood

Comment (aladdin-add) [reply docs/shared-settings.md:67]: @scagood your insights are wanted here. :)

Comment (scagood) [reply docs/shared-settings.md:67]: :thinking: I am on the fence here, because `enhanced-resolve` covers typescripts silly resolution options far better than `import.meta.resolve` -- I feel like we may need to re-write large chunks of the resolver to make it work.

That being said, I am happy to leave this as documenting only `resolverConfig.modules`, but keeping in the other options as undocumented or possibly even state they're unsupported options.

Comment (aladdin-add) [reply docs/shared-settings.md:67]: friendly ping @GertSallaerts. are you still working on this?

Comment (GertSallaerts) [reply docs/shared-settings.md]: Updated the docs, what do you think? It's been a busy week on my end too 😅

**Review by aladdin-add — APPROVED** [review]
LGTM, thanks!
